src/atomate2/abinit/sets/response.py

```diff
@@ -0,0 +1,66 @@
+"""Module defining response function Abinit input set generators."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Callable
+
+from abipy.abio.factories import (
+    ddepert_from_gsinput,
+    ddkpert_from_gsinput,
+    dtepert_from_gsinput,
+)
+from abipy.abio.input_tags import DDE, DDK, DTE, MOLECULAR_DYNAMICS, RELAX, SCF
+
+if TYPE_CHECKING:
+    from pymatgen.io.abinit import PseudoTable
+
+from atomate2.abinit.sets.core import NonSCFSetGenerator, StaticSetGenerator
+
+__all__ = [
+    "DdkSetGenerator",
+    "DdeSetGenerator",
+    "DteSetGenerator",
+]
+
+
+GS_RESTART_FROM_DEPS: tuple = (f"{SCF}|{RELAX}|{MOLECULAR_DYNAMICS}:WFK|DEN",)
+
+
+@dataclass
+class DdkSetGenerator(NonSCFSetGenerator):
+    """Class to generate Abinit DDK input sets."""
+
+    calc_type: str = "DDK"
+    factory: Callable = ddkpert_from_gsinput
+    restart_from_deps: tuple = (f"{DDK}:1WF",)
+    prev_outputs_deps: tuple = (f"{SCF}:WFK",)
+    nbands_factor: float = 1.0  # TODO: to decide if nbdbuf or not
```

Comment:
The number of bands is something we wanted to investigate because including a buffer really speeds up the convergence of the calculations, but then we noticed that the result was different... in doubt we should only use the valence bands I think.
For the ```NonSCFSetGenerator``` or ```StaticSetGenerator``` inheritance, it is just because the DDK step should be an NSCF calculation while the DDE should be a SCF one... I think... not entirely sure though.

Reply:
I would not get in the matter of the procedure, but rather on how it is planned to be used. As far as I can see the main difference between the two is that `NonSCFSetGenerator` defines the `nbands_factor`. The rest is basically overridden. It could basically even be a directly a subclass of `AbinitInputGenerator` and it will not change much. The main point is if the `nbands_factor` is needed in the end or not. If not I would probably switch to another base class.

On a side note, I almost always used a small buffer of bands. I did not see differences, but I admit I absolutely did not test this in depth. How much of a difference does this make?

Reply:
The use of a buffer and more bands than the VBs is still smth I have to investigate... it was first noticed by a student who was computing the static refractive index... I don't remember how different the results were though... in any case I think we should keep the option of being able to modify nband and nbdbuf no?

Reply:
From what I understand from ```DTEFWWorkflow``` of Abiflows, the factory for the static calculation is ```scf_for_phonons```, meaning that there is a buffer (nband = VBs+nbdbuf which is 4), but then ```nband=VBs``` and ```nbdbuf=None``` for the DDK, DDE and DTE via ```dte_from_gsinput```.

Reply:
> From what I understand from `DTEFWWorkflow` of Abiflows, the factory for the static calculation is `scf_for_phonons`, meaning that there is a buffer (nband = VBs+nbdbuf which is 4), but then `nband=VBs` and `nbdbuf=None` for the DDK, DDE and DTE via `dte_from_gsinput`.

Indeed, now that you tell me, if I remember correctly the dte part only works if the number of bands match the number of electron in the valence. So the factory was doing this for all the perturbations. I did not think of adding the option for more bands in the intermediate steps.

Let's leave it like this then.

Reply:
I think it is still worth investigating because adding a buffer for the intermediate steps can help the convergence, but we should be sure that it does not modify the results... at the moment only the DTE has nband=VBs... so it might be nice to be sure that the results are okay when nband>VBs for the other perturbations... also because I already used this workflow to compute the SHG for ~200 materials...